pkg/agents/team4/finalAgent/agent.go

```diff
@@ -0,0 +1,197 @@
+
+	globalTrust      float32            // the overall trust the agent has in other agents in the tower
+	coefficients     []float32          // the amount trust score changes by for certain actions
+	lastFoodTaken    food.FoodType      // food taken on the previous day
+	sentMessages     []messages.Message // TODO: make it a map hashed by messageIDs
+	responseMessages []messages.Message // TODO: make it a map hashed by messageIDs
+	lastPlatFood     food.FoodType      // last seen food on the platform
+	maxFoodLimit     food.FoodType      // maximum food we want to allow others to eat
+	messageCounter   int                // the total number of messages we send in a day
+	globalTrustLimit float32            // limit to check whether to be selfish or not
+	lastAge          int                // the age of the agent on the previous day
+	healthStatus     int
+}
+
+type CustomAgentEvo struct {
+	*infra.Base
+	// new params
+	params CustomAgentEvoParams
+}
+
+// Data struct to hold evo param values
+type LoadedData struct {
+	FoodToEat  []int
+	DaysToWait []int
+}
+
+func InitaliseParams(baseAgent *infra.Base) CustomAgentEvoParams {
+
+	foodToEat := map[string][]int{
+		"selfish":  {baseAgent.HealthInfo().HPReqCToW, 7, 9, 11}, // TODO: to optimise more baseAgent.HealthInfo().HPReqCToW
+		"selfless": {baseAgent.HealthInfo().HPReqCToW, 7, 9, 11}, // TODO: to optimise more
+	}
+	daysToWait := map[string][]int{
+		"selfish":  {0, 1, 4, 3}, // TODO: to optimise more int(baseAgent.HealthInfo().MaxDayCritical / 2)
+		"selfless": {0, 1, 4, 3}, // TODO: to optimise more
+	}
+
+	return CustomAgentEvoParams{ //initialise the parameters of the agent
+		locked:             true,
+		foodToEat:          foodToEat,
+		daysToWait:         daysToWait,
+		currentPersonality: "selfless",
+		ageLastEaten:       0,
+		morality:           100 * rand.Float64(), // TODO: Use this properly
+		traumaScaleFactor:  1,
+		healthStatus:       3,
+		globalTrust:        0.0,
+		coefficients:       []float32{2, 4, 8}, // TODO: maybe train these co-efficients using evolutionary algorithm
+		lastFoodTaken:      0,
+		sentMessages:       []messages.Message{},
+		responseMessages:   []messages.Message{},
+		lastPlatFood:       -1,
+		maxFoodLimit:       50,
+		messageCounter:     0,
+		globalTrustLimit:   75,
+		lastAge:            0,
+	}
+}
+
+func New(baseAgent *infra.Base) (infra.Agent, error) {
+	//create other parameters
+	return &CustomAgentEvo{
+		Base:   baseAgent,
+		params: InitaliseParams(baseAgent),
+	}, nil
+}
+
+// Checks if neighbour below has eaten
+func (a *CustomAgentEvo) NeighbourFoodEaten() food.FoodType {
+	if a.CurrPlatFood() != -1 {
+		if !a.PlatformOnFloor() && a.CurrPlatFood() != a.params.lastPlatFood {
+			return a.params.lastPlatFood - a.CurrPlatFood()
+		}
+		return 0
+	}
+	return -1
+}
+
+// removes a specific message from a message array
+func remove(slice []messages.Message, s int) []messages.Message {
+	return append(slice[:s], slice[s+1:]...)
```

Comment:
it is never called for an invalid index since it is called inside of a for loop.

Reply:
should have made that clear somewhere probably

Reply:
Cool. When I'm reviewing I don't look at the function references (for most of them), just at the function implementation :) otherwise it's a huge cognitive load. Also, if the function was to be used for more complex indexing in the future (which could fail or have edge cases), implementing a check (and maybe throwing an out-of-bounds error) is always safer than getting a seg fault